chat/consumers.py

```diff
@@ -114,91 +195,99 @@ async def receive(self, text_data):
         2. Retrieves the Room by `room_name` or creates it if it doesn't exist.
-           Uses MongoEngine upsert to reduce race conditions.
-        3. Ensures the user is a participant of the room.
-           For large participant lists, consider using a set or indexed field for efficiency.
-        4. Creates and saves a Message object linked to the room and user.
-        5. Logs all actions for auditability, truncating long messages for readability.
+        3. Ensures the sender ID is a participant of the room.
+        4. Creates and saves a Message object linked to the room.
+        5. Logs actions for auditability, truncating long messages for readability.
 
         Notes:
-        - Multiple parallel requests creating the same room could cause race conditions.
-          Upsert or MongoDB transactions help prevent duplicates.
-        - Ensure Room and Message models’ required fields and signals (timestamps, hooks) are respected.
+        - `receiver_id` can be None for group messages.
+        - Race conditions can occur if multiple requests create the same room; consider using upsert or transactions.
         """
         user = self.scope["user"]
         if not user:
             logger.warning("Anonymous user attempted to send message: %s", message)
             raise ValueError("Anonymous users cannot send messages")
 
-        room, created = Room.objects.get_or_create(name=room_name)
-
-        if created:
-            room.participants = [user]
-            try:
-                room.save()
-                logger.info("Created new room '%s' and added user %s", room_name, user.email)
-            except ValidationError as ve:
-                logger.error("Room creation failed: %s", ve)
-                raise ve
-        else:
-            participant_ids = {u.id for u in room.participants}
-            if user.id not in participant_ids:
-                room.participants.append(user)
+        try:
+            room = Room.objects.get(name=room_name)
+            created = False
+        except DoesNotExist:
+            room = Room(name=room_name, participants=[str(sender_id)])
+            room.save()
+            created = True
+
+        if not created:
+            if str(sender_id) not in room.participants:
+                room.participants.append(str(sender_id))
                 try:
                     room.save()
-                    logger.info("Added user %s to existing room '%s'", user.email, room_name)
+                    logger.info("Added user %s to existing room '%s'", sender_id, room_name)
                 except ValidationError as ve:
                     logger.error("Adding user to room failed: %s", ve)
                     raise ve
 
-        msg = Message(room=room, sender=user, text=message)
+        msg = Message(
+            room=room,
+            sender_id=str(sender_id),
+            receiver_id=str(receiver_id) if receiver_id else None,
```

Comment:
Participants are stored as strings, but receiver_id is allowed to be None, which may cause inconsistent types in participants. Always ensure receiver_id is a string or exclude it.
```
receiver_id = str(receiver_id) if receiver_id is not None else None
```